File: app/routes/camera.py
from flask import Blueprint, request, jsonify, make_response
import base64
import cv2
import numpy as np
import time

from src.state import get_state_collector
from app.services.camera_service import CameraService
from app.config import config

camera_bp = Blueprint("camera", __name__, url_prefix="/api/camera")


@camera_bp.route("/snapshot")
def snapshot():
    """返回 320x240 图片 + 小车自己的 m 和 c

    Returns:
        {"image": "<base64 jpeg>", "width": 320, "height": 240,
         "m": 3062.0, "c": -5.0}
    """
    service = CameraService.get_instance()
    service._ensure_camera()
    ret, frame = service.read()
    if not ret or frame is None:
        return jsonify({"error": "camera not available"}), 500

    # letterbox: 等比缩放 + 黑边填充到 320x240，保证画面不变形
    H, W = frame.shape[:2]
    r = min(320.0 / W, 240.0 / H)
    new_w, new_h = int(round(W * r)), int(round(H * r))
    resized = cv2.resize(frame, (new_w, new_h), interpolation=cv2.INTER_LINEAR)
    canvas = np.zeros((240, 320, 3), dtype=np.uint8)
    dw = (320 - new_w) // 2
    dh = (240 - new_h) // 2
    canvas[dh:dh + new_h, dw:dw + new_w] = resized
    frame_out = canvas
    ret, jpg = cv2.imencode(".jpg", frame_out, [cv2.IMWRITE_JPEG_QUALITY, 70])
    if not ret:
        return jsonify({"error": "encode failed"}), 500

    return jsonify({
        "image": base64.b64encode(jpg.tobytes()).decode("ascii"),
        "width": 320,
        "height": 240,
        "format": "jpeg",
        "m": config.calib_m,
        "c": config.calib_c,
    })


@camera_bp.route("/open", methods=["POST"])
def camera_open():
    service = CameraService.get_instance()
    service._ensure_camera()
    available = service.is_available()
    return jsonify({"camera_on": available})


@camera_bp.route("/close", methods=["POST"])
def camera_close():
    CameraService.get_instance().close()
    return jsonify({"camera_on": False})


@camera_bp.route("/status")
def camera_status():
    available = CameraService.get_instance().is_available()
    return jsonify({"camera_on": available})


@camera_bp.route("/stream")
def video_stream():
    """MJPEG视频流 - 直接从Camera读取最新帧，无竞争"""
    collector = get_state_collector()

    if collector._camera is None or not hasattr(collector._camera, '_frame_ready'):
        return jsonify({"error": "camera not available"}), 500

    def generate():
        try:
            while True:
                if not collector._running:
                    break
                # 直接从 Camera 读最新帧引用，无 event 竞争
                ret, frame = collector._camera.read()
                if not ret or frame is None:
                    time.sleep(0.05)
                    continue
                ret, jpg = cv2.imencode('.jpg', frame, [cv2.IMWRITE_JPEG_QUALITY, 40])
                if not ret:
                    continue
                yield (b'--frame\r\n'
                       b'Content-Type: image/jpeg\r\n\r\n' + jpg.tobytes() + b'\r\n')
                time.sleep(0.03)  # ~30fps 上限
        except GeneratorExit:
            pass

    response = make_response(generate())
    response.headers['Content-Type'] = 'multipart/x-mixed-replace; boundary=frame'
    response.headers['Access-Control-Allow-Origin'] = '*'
    response.headers['Cache-Control'] = 'no-cache, no-store, must-revalidate'
    response.headers['Pragma'] = 'no-cache'
    response.headers['Expires'] = '0'
    return response


@camera_bp.route("/speed")
def speed_status():
    """只获取左右轮速度和目标值"""
    collector = get_state_collector()
    status = collector.get_status()
    return jsonify({
        "left_speed": status.left_speed,
        "right_speed": status.right_speed,
        "left_target": status.left_target,
        "right_target": status.right_target,
        "gripper_status": status.gripper_status,
        "gripper_target": status.gripper_target,
        "timestamp_ms": status.timestamp_ms,
    })


@camera_bp.route("/all_status")
def all_status():
    collector = get_state_collector()
    status = collector.get_status()
    image = collector.get_image()

    image_data = None
    if image is not None:
        ret, jpg = cv2.imencode('.jpg', image, [cv2.IMWRITE_JPEG_QUALITY, 25])
        if ret:
            image_data = base64.b64encode(jpg.tobytes()).decode('ascii')

    return jsonify({
        "timestamp": request.args.get("timestamp"),
        "left_speed": status.left_speed,
        "right_speed": status.right_speed,
        "left_target": status.left_target,
        "right_target": status.right_target,
        "gripper_status": status.gripper_status,
        "gripper_target": status.gripper_target,
        "timestamp_ms": status.timestamp_ms,
        "image": image_data,
        "image_format": "jpeg",
    })

File: cpp/capp/src/routes/camera.cpp
// 摄像头
//
// 入口：register_camera_routes()（由 src/routes.cpp 的 register_routes 调用）
//
// 由 capp/src/routes.cpp 按域拆出来（对照 app/routes/*.py 的分法）。
// 对应 app/routes/camera.py。

#include "routes_internal.hpp"

#include "csrc/base64.hpp"
#include "csrc/log.hpp"
#include <cstdlib>
#include <thread>

namespace capp {
namespace routes {

// ── 摄像头 ──

void register_camera_routes(Router& router, AppContext& ctx) {


    // ── /api/camera ──
    // 注：屏显示跟随摄像头开关，但对前端透明 ——
    //     open 后台自动出图、close 后台自动清屏，接口不暴露屏状态。
    router.add("GET", "/api/camera/status", [&ctx](const HttpRequest&, HttpResponse& resp, ClientConn&, AppContext&) {
        Json j;
        j["camera_on"] = ctx.camera_on && ctx.camera.is_available();
        resp.set_json(j);
    });

    router.add("POST", "/api/camera/open", [&ctx](const HttpRequest&, HttpResponse& resp, ClientConn&, AppContext&) {
        bool ok = ensure_camera(ctx);   // 摄像头打开 → 屏随之开始显示（对前端透明）
        Json j;
        j["camera_on"] = ok && ctx.camera.is_available();
        resp.set_json(j, ok ? 200 : 500);
    });

    router.add("POST", "/api/camera/close", [&ctx](const HttpRequest&, HttpResponse& resp, ClientConn&, AppContext&) {
        close_camera(ctx);              // 摄像头关闭 → 屏清屏熄灭（对前端透明）
        Json j;
        j["camera_on"] = false;
        resp.set_json(j);
    });

    router.add("GET", "/api/camera/snapshot", [&ctx](const HttpRequest&, HttpResponse& resp, ClientConn&, AppContext&) {
        ensure_camera(ctx);
        std::vector<uint8_t> jpeg;
        if (!current_jpeg(ctx, 70, jpeg)) {
            resp.set_error("camera not available", 500);
            return;
        }
        int w = 0, h = 0;
        csrc::Camera::jpeg_get_size(jpeg.data(), jpeg.size(), w, h);
        Json j;
        j["image"] = csrc::base64_encode(jpeg.data(), jpeg.size());
        j["width"] = csrc::Json((int64_t)w);
        j["height"] = csrc::Json((int64_t)h);
        j["format"] = "jpeg";
        j["m"] = ctx.config.calib_m;
        j["c"] = ctx.config.calib_c;
        resp.set_json(j);
    });

    // MJPEG 流（Python /api/camera/stream 契约，支持 ?fps=N 覆盖，默认 15fps）
    router.add("GET", "/api/camera/stream", [&ctx](const HttpRequest& req, HttpResponse& resp, ClientConn& conn, AppContext&) {
        resp.stream = true;
        ensure_camera(ctx);
        int fps = 15;
        std::string fps_s = req.query_param("fps");
        if (!fps_s.empty()) { fps = atoi(fps_s.c_str()); if (fps < 1) fps = 1; if (fps > 30) fps = 30; }
        auto min_interval = std::chrono::milliseconds(1000 / fps);

        std::string head =
            "HTTP/1.1 200 OK\r\n"
            "Content-Type: multipart/x-mixed-replace; boundary=frame\r\n"
            "Access-Control-Allow-Origin: *\r\n"
            "Cache-Control: no-cache, no-store, must-revalidate\r\n"
            "Pragma: no-cache\r\n"
            "Expires: 0\r\n"
            "Connection: close\r\n\r\n";
        conn.write_all(head);

        uint64_t last_ts = 0;
        auto last_send = std::chrono::steady_clock::now();
        // stream_scale 配置开关 + 尺寸>0 → 服务端缩放重编码后下发（省 WiFi 带宽）；
        // 关闭或尺寸无效 → 直通原帧。
        const bool downscale = ctx.config.camera.stream_scale &&
                               ctx.config.camera.stream_width > 0 &&
                               ctx.config.camera.stream_height > 0;
        // 编码输出缓冲跨帧复用，避免每帧 malloc（长时间运行更稳）
        std::vector<uint8_t> jpeg;
        // 有浏览器在看流 → 屏显示降帧（[display] fps_streaming，0=暂停）：
        // 单核 SoC 上"显示 + 取流"会 CPU 饱和，拖慢网页看摄像头的帧率/延迟。
        ctx.display.set_streaming(true);
        while (true) {
            auto now = std::chrono::steady_clock::now();
            if (now - last_send >= min_interval) {
                bool ok = false;
                uint64_t ts = 0;
                auto t0 = std::chrono::steady_clock::now();
                if (downscale) {
                    // 缩放路径：解码走 Camera::latest_rgb **共享缓存** ——
                    // 屏显示线程通常已解过这一帧，这里直接命中，省掉整帧解码
                    // （640x360 MJPEG → 320 宽，C906 上约 10ms/帧）。
                    csrc::Camera::RgbFrame rgb;
                    if (ctx.camera.latest_rgb(ctx.config.camera.stream_width, rgb) &&
                        !rgb.data.empty() && rgb.ts_ms != last_ts) {
                        jpeg.clear();
                        ok = build_stream_jpeg_rgb(ctx, rgb, jpeg);
                        ts = rgb.ts_ms;
                    }
                } else {
                    // 直通路径：原帧直接下发，完全不碰解码
                    csrc::Camera::Frame f;
                    if (ctx.camera.read_latest(f) && !f.data.empty() && f.ts_ms != last_ts &&
                        csrc::Camera::is_jpeg(f.data.data(), f.data.size())) {
                        jpeg = std::move(f.data);
                        ok = true;
                        ts = f.ts_ms;
                    }
                }
                double enc_ms = std::chrono::duration<double, std::milli>(
                                    std::chrono::steady_clock::now() - t0).count();
                // 诊断：单帧耗时 >120ms 即肉眼可见卡顿，记录一次(每帧, debug 级)
                if (enc_ms > 120.0) {
                    CAM_DEBUG("camera stream frame encode %.0fms (cpu busy? size=%zu)",
                              enc_ms, jpeg.size());
                }
                if (ok && !jpeg.empty()) {
                    // MJPEG 直通/重编码帧：头 + jpeg + 尾拼成一个 buffer 一次 write（减少系统调用）
                    std::string part = "--frame\r\nContent-Type: image/jpeg\r\n"
                                       "Content-Length: " + std::to_string(jpeg.size()) +
                                       "\r\n\r\n";
                    std::string out;
                    out.reserve(part.size() + jpeg.size() + 2);
                    out += part;
                    out.append((const char*)jpeg.data(), jpeg.size());
                    out += "\r\n";
                    if (!conn.write_all(out)) break;
                    last_ts = ts;
                    last_send = now;
                }
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(10));
        }
        ctx.display.set_streaming(false);   // 没人看流了 → 屏恢复 fps
        conn.close();
    });

    router.add("GET", "/api/camera/speed", [&ctx](const HttpRequest&, HttpResponse& resp, ClientConn&, AppContext&) {
        csrc::RobotStatus s = ctx.collector.get_status();
        Json j;
        j["left_speed"] = s.left_speed;
        j["right_speed"] = s.right_speed;
        j["left_target"] = csrc::Json((int64_t)s.left_target);
        j["right_target"] = csrc::Json((int64_t)s.right_target);
        j["gripper_status"] = s.gripper_status;
        j["gripper_target"] = csrc::Json((int64_t)s.gripper_target);
        j["timestamp_ms"] = csrc::Json((int64_t)s.timestamp_ms);
        resp.set_json(j);
    });

    router.add("GET", "/api/camera/all_status", [&ctx](const HttpRequest& req, HttpResponse& resp, ClientConn&, AppContext&) {
        csrc::RobotStatus s = ctx.collector.get_status();
        Json j;
        j["timestamp"] = req.query_param("timestamp");
        j["left_speed"] = s.left_speed;
        j["right_speed"] = s.right_speed;
        j["left_target"] = csrc::Json((int64_t)s.left_target);
        j["right_target"] = csrc::Json((int64_t)s.right_target);
        j["gripper_status"] = s.gripper_status;
        j["gripper_target"] = csrc::Json((int64_t)s.gripper_target);
        j["timestamp_ms"] = csrc::Json((int64_t)s.timestamp_ms);
        j["motor"] = motor_status_json(ctx);

        std::vector<uint8_t> jpeg;
        if (current_jpeg(ctx, 25, jpeg) && !jpeg.empty()) {
            j["image"] = csrc::base64_encode(jpeg.data(), jpeg.size());
        } else {
            j["image"] = Json();
        }
        j["image_format"] = "jpeg";
        resp.set_json(j);
    });
}
}  // namespace routes
}  // namespace capp
